---
output: 
  html_document:
params: 
  genename: "smc1a"
  normalization: "rlog"

  experiment_name: "jcoffman008"
  

  
  filepath_Z_normalized: "jcoffman_008_merged_nogeneduplicates_Z_normalized.txt"
  filepath_loading_scores: "jcoffman_008_merged_nogeneduplicates_meaningful_pc_loading_scores.txt"
  filepath_pca_eigenvalues: "jcoffman_008_merged_nogeneduplicates_pca_eigenvalues.txt"
  filepath_design_meaningful: "jcoffman_008_merged_nogeneduplicates_design_meaningful.txt"

  filepath_rlog: "jcoffman_008.klf9_multiTime_2021_merged_nogeneduplicates.rlog.tsv"
  filepath_deseq: "jcoffman_008.klf9_multiTime_2021_merged_nogeneduplicates_PC1_time.DESeq2_out.tsv"
  filepath_deseq3hb: "jcoffman_008.klf9_multiTime_2021_merged_3hb_nogeneduplicates.DESeq2_out.tsv"
  filepath_deseqlightson: "jcoffman_008.klf9_multiTime_2021_merged_lightson_nogeneduplicates.DESeq2_out.tsv"
  filepath_deseq2.5ha: "jcoffman_008.klf9_multiTime_2021_merged_2.5ha_nogeneduplicates.DESeq2_out.tsv"



title: "Summary of `r params$genename` behavior in jcoffman 008"
classoption: landscape

---


```{r setup,echo=FALSE,include=FALSE}

#USAGE: enter terminal command in format R -e "rmarkdown::render('jcoffman008genesummary.Rmd',params=list(genename='lss',normalization=zscore))"
## the optional znorm parameter allows the user to specify if z-normalization is preferred over rlog normalization for representing expression 

library(tidyr)
library(gridExtra)
library(ggplot2)
library(reshape2)
library(stringr)
library(knitr)

######Saving useful output files as dataframes. Creates objects nonspecific to the gene


#rlog normalized counts of genes (from DESEQ output) this transformation takes into account variation of counts between samples reduces noise.
rlog <- read.delim(file= params$filepath_rlog)
#z-score (normalization that assumes normal distribution) from PCA 
zscore<- read.delim(file= params$filepath_Z_normalized)
##storing PCA gene loadings from the run with all samples
pca<- read.delim(file= params$filepath_loading_scores)
all_eigenvalues <- read.delim(file=params$filepath_pca_eigenvalues)
all_design<- read.delim(file=params$filepath_design_meaningful)
all_design$replicate <- as.factor(all_design$replicate)
all_design$time<- as.factor(all_design$time)

#storing DESEQ results
deseq <- read.delim(file= params$filepath_deseq)
dsq.3hb <- read.delim(file=params$filepath_deseq3hb)
dsq.lightson<- read.delim(file= params$filepath_deseqlightson)
dsq.2.5ha<- read.delim(file= params$filepath_deseq2.5ha)
#getting PCA results in format for plot 
long_pca<- melt(pca[c(1,2,3,4,5)])

#getting DESEQ results in format for plot 
dsq.3hb<- dsq.3hb[,c("X","baseMean","log2FoldChange","padj")]
dsq.lightson<- dsq.lightson[,c("X","baseMean","log2FoldChange","padj")]
dsq.2.5ha<- dsq.2.5ha[,c("X","baseMean","log2FoldChange","padj")]
names(dsq.3hb)[names(dsq.3hb) == "baseMean"] <- "baseMean3hb"
names(dsq.3hb)[names(dsq.3hb) == "log2FoldChange"] <- "3hb"
names(dsq.3hb)[names(dsq.3hb) == "padj"] <- "padj3hb"
names(dsq.lightson)[names(dsq.lightson) == "baseMean"] <- "baseMeanlightson"
names(dsq.lightson)[names(dsq.lightson) == "log2FoldChange"] <- "lightson"
names(dsq.lightson)[names(dsq.lightson) == "padj"] <- "padjlightson"
names(dsq.2.5ha)[names(dsq.2.5ha) == "baseMean"] <- "baseMean2.5ha"
names(dsq.2.5ha)[names(dsq.2.5ha) == "log2FoldChange"] <- "2.5ha"
names(dsq.2.5ha)[names(dsq.2.5ha) == "padj"] <- "padj2.5ha"
dsq_fc.merge1 <-merge(dsq.3hb[,c("X","3hb")],dsq.lightson[,c("X","lightson")])
dsq_fc.merge2<- merge(dsq_fc.merge1,dsq.2.5ha[,c("X","2.5ha")])
long_dsq_fc<- melt(dsq_fc.merge2)
names(long_dsq_fc)[names(long_dsq_fc)=="variable"]<- "time"
names(long_dsq_fc)[names(long_dsq_fc)=="value"]<- "log2FoldChange"
dsq_padj.merge1<- merge(dsq.3hb[,c("X","padj3hb")],dsq.lightson[,c("X","padjlightson")])
dsq_padj.merge2<- merge(dsq_padj.merge1,dsq.2.5ha[,c("X","padj2.5ha")])
long_dsq_padj<- melt(dsq_padj.merge2)
names(long_dsq_padj)[names(long_dsq_padj)=="variable"]<- "time"
long_dsq_padj$time<- str_replace_all(long_dsq_padj$time,"padj","")
names(long_dsq_padj)[names(long_dsq_padj)=="value"]<- "padj"
dsq_mean.merge1<- merge(dsq.3hb[,c("X","baseMean3hb")],dsq.lightson[,c("X","baseMeanlightson")])
dsq_mean.merge2<- merge(dsq_mean.merge1,dsq.2.5ha[,c("X","baseMean2.5ha")])
long_dsq_mean<- melt(dsq_mean.merge2)
names(long_dsq_mean)[names(long_dsq_mean)=="variable"]<- "time"
long_dsq_mean$time<- str_replace_all(long_dsq_mean$time,"baseMean","")
names(long_dsq_mean)[names(long_dsq_mean)=="value"]<- "baseMean"
long_dsq.merge1<- merge(long_dsq_fc,long_dsq_padj,by = intersect(names(long_dsq_fc), names(long_dsq_padj)))
long_dsq<-merge(long_dsq.merge1,long_dsq_mean, by = intersect(names(long_dsq_fc), names(long_dsq_padj)))

##Separating the labels of file-based dataframes
rlog<- separate(data = rlog, col = X, into = c("GeneID", "GeneName"), sep = "\\_")
zscore<- separate(data = zscore, col = X, into = c("GeneID", "GeneName"), sep = "\\_")
pca<- separate(data = pca, col = X, into = c("GeneID", "GeneName"), sep = "\\_")
deseq<- separate(data = deseq, col = X, into = c("GeneID", "GeneName"), sep = "\\_")
dsq.3hb<- separate(data = dsq.3hb, col = X, into = c("GeneID", "GeneName"), sep = "\\_")
dsq.lightson<- separate(data = dsq.lightson, col = X, into = c("GeneID", "GeneName"), sep = "\\_")
dsq.2.5ha<- separate(data = dsq.2.5ha, col = X, into = c("GeneID", "GeneName"), sep = "\\_")
long_pca<- separate(data = long_pca, col = X, into = c("GeneID", "GeneName"), sep = "\\_")
long_dsq<- separate(data = long_dsq, col = X, into = c("GeneID", "GeneName"), sep = "\\_")


#these lines are what allow WT to come before KO in plots 
all_design$genetics<- factor(all_design$genetics,levels=c("WT","KO"))
all_design$X<- factor(all_design$X,levels=all_design$X[order(all_design$genetics)])

genename<- params$genename
#controlled error message if the gene is not in the output files that the plots use
if(genename %in% long_dsq$GeneName==FALSE | genename %in% long_pca$GeneName==FALSE){
  stop(paste("\n\n\n ********* WHY THE CODE FAILED!!: You entered the following gene name:  \n\t\t" ,genename, "\nThis is not a recognized gene name or it is insufficiently expressed in experiment (did not pass the DESEQ or PCA cutoffs). Please check capitalization or enter another gene.*********\n\n\n",sep=""))
}

#adding the rlog and zscore normalized counts of this gene from each of these samples to the design file
design_genespecific<- all_design
design_genespecific$target_gene_rlog<- as.numeric(t(subset(rlog,GeneName==genename))[c(3:ncol(rlog)),])
design_genespecific$target_gene_zscore<- as.numeric(t(subset(zscore,GeneName==genename))[c(3:ncol(zscore)),])


```
# {.tabset}




## PCA Results
The information on Principal Components below contextualizes the `r genename`-specific output. As the facet plots demonstrate, PC1 captures the replicate effect, PC2 and PC4 capture circadian oscillations in gene expression, and PC3 captures the effect of the *klf9*-knockout on gene expression. 

```{r generatefacetplots, out.width="150%", warning=FALSE, echo=FALSE, results='hide'}
###Facet plots to provide meaning to Principal Components. Only has to generate the plots if this is the first time running the script  

facetplotpng<- paste(params$experiment_name,"PCAfacetplots.png",sep="")

png(filename=facetplotpng, width=600, height=400)

if(!file.exists(facetplotpng)){
install.packages("patchwork",repos = "http://cran.us.r-project.org")
library(patchwork)
#creating labels that allow plots to say the percent variance of a PC
all_PC1_label<- paste("PC1: ", round(all_eigenvalues[1,]$variance.percent,2), "% of variance ",sep="")
all_PC2_label<- paste("PC2: ", round(all_eigenvalues[2,]$variance.percent,2), "% of variance ",sep="")
all_PC3_label<- paste("PC3: ", round(all_eigenvalues[3,]$variance.percent,2), "% of variance ",sep="")
all_PC4_label<- paste("PC4: ", round(all_eigenvalues[4,]$variance.percent,2), "% of variance ",sep="")
#facet plot of PC1
facetplotPC1 <- ggplot(data=all_design, aes(x=X, y=PC1,fill=genetics))+geom_bar(stat="identity", position=position_dodge())+ scale_fill_manual(values = c("black", "red")) +
  facet_grid(. ~ time, scales='free',labeller=label_both)+
  labs(y="Sample Loading")+
  theme_bw()+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(), panel.border = element_blank(), panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())+ 
  ggtitle(all_PC1_label)
#facet plot of PC2
facetplotPC2 <-ggplot(data=all_design, aes(x=X, y=PC2,fill=genetics))+geom_bar(stat="identity", position=position_dodge())+ scale_fill_manual(values = c("black", "red")) +
  facet_grid(. ~ time, scales='free',labeller=label_both)+
  theme_bw()+
  labs(y="Sample Loading")+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(), panel.border = element_blank(), panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())+ 
  ggtitle(all_PC2_label)
#facet plot of PC3
facetplotPC3 <-ggplot(data=all_design, aes(x=X, y=PC3,fill=genetics))+geom_bar(stat="identity", position=position_dodge())+ scale_fill_manual(values = c("black", "red")) +
  facet_grid(. ~ time, scales='free',labeller=label_both)+
  theme_bw()+
  labs(y="Sample Loading")+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(), panel.border = element_blank(), panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())+ 
  ggtitle(all_PC3_label)
#facet plot of PC4
facetplotPC4 <-ggplot(data=all_design, aes(x=X, y=PC4,fill=genetics))+geom_bar(stat="identity", position=position_dodge())+ scale_fill_manual(values = c("black", "red")) +
  facet_grid(. ~ time, scales='free',labeller=label_both)+
  labs(y="Sample Loading")+
  theme_bw()+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(), panel.border = element_blank(), panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())+ 
  ggtitle(all_PC4_label)

  facetplotPC1 + facetplotPC2 + facetplotPC3 + facetplotPC4
  }else{
  include_graphics(facetplotpng)
  }
  dev.off()

``` 
``` {r displayfacetplots, out.width="150%", warning=FALSE, echo=FALSE}
include_graphics(facetplotpng)

```

### `r genename` in PCA results

```{r PCAgenespecific, echo=FALSE,out.width=c('50%', '50%'), fig.show='hold'}

###Creating plot to show relative PC loadings
ggplot(data=long_pca, aes(x=variable,y=value))+geom_violin()+
  geom_point(data=subset(long_pca,GeneName==genename), color= "green",aes(y=value), size=4)+ theme(legend.position = "none", aspect.ratio = 1.5,panel.background = element_blank())+
  geom_hline(yintercept=0,linetype="dashed")+
  labs(x="",y= "Gene Loading")+
  ggtitle(paste("Relative ",genename," loadings in PCA of all samples",sep=""))

#examine replicate effect: plot sample's PC1 loading vs. target gene expression level (zscore norm htseq count)
#if the gene is associated with the replicate effect it will have a x=y or x=-y distribution of samples
ggplot(design_genespecific, aes(x=PC1, y=target_gene_zscore, label=replicate,color=genetics,pch=time))+geom_point(size=4) +scale_color_manual(values = c("black", "red")) +
  geom_vline(xintercept=0,linetype="dashed")+ geom_hline(yintercept=0,linetype="dashed")+
  labs(x="PC1 loading of sample", y= paste("zscore-norm count of ", genename, " in sample",sep=""))+
  ggtitle(paste("Is ", genename, " expression associated with the replicate effect?",sep=""))+ theme( aspect.ratio = 1,panel.background = element_blank())

PC1_percentile <- round(mean(pca$PC1 < subset(pca,GeneName==genename)$PC1 ) ,2)
PC2_percentile <- round(mean(pca$PC2 < subset(pca,GeneName==genename)$PC2 ) ,2)
PC3_percentile <- round(mean(pca$PC3 < subset(pca,GeneName==genename)$PC3 ) ,2)
PC4_percentile <- round(mean(pca$PC4 < subset(pca,GeneName==genename)$PC4 ) ,2)
```

Compared to the distribution of loadings for all genes examined, `r genename` is loaded in the  ```r PC1_percentile``` percentile for PC1, ```r PC2_percentile``` percentile for PC2,```r PC3_percentile``` percentile for PC3, and ```r PC4_percentile``` percentile for PC4. 

While it is known that the replicate effect is the dominating source of variance in this experiment, one can infer if `r genename` follows the expression pattern between replicates captured by PC1 with the plot (above right) of all samples' `r genename` expression levels against their PC1 loadings. If the samples organize in a linear manner,  `r genename` expression is affected by the replicate effect described by PC1. The time and genetics of the samples are provided as supplementary information. 



## `r genename` expression

```{r expression,echo=FALSE,warning=FALSE}


if(params$normalization=="zscore"){
  ggplot(data=design_genespecific, aes(x=X, y=target_gene_zscore,color=genetics))+geom_point(size=5)+scale_color_manual(values = c("black","red")) + 
  facet_grid(. ~ time, scales='free',labeller=label_both,switch="x")+
  geom_text(aes(label = replicate), vjust = -.2, size=5,color="blue")+
  theme_bw()+
  labs(x="Sample", y= "zscore norm count")+
  theme(aspect.ratio=2.5,
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(), panel.border = element_blank(), panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())+ 
  ggtitle(paste("Z-score normalized ",genename, " counts across samples", sep=""))

}else{
  ggplot(data=design_genespecific, aes(x=X, y=target_gene_rlog,color=genetics))+geom_point(size=5)+scale_color_manual(values = c("black","red")) + 
  facet_grid(. ~ time, scales='free',labeller=label_both,switch="x")+
  geom_text(aes(label = replicate), vjust = -.2, size=5,color="blue")+
  theme_bw()+
  labs(x="Sample", y= "rlog norm count")+
  theme(aspect.ratio=2.5,
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(), panel.border = element_blank(), panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())+ 
  ggtitle(paste("rlog normalized ",genename, " counts across samples", sep=""))

}


```

## Effect of klf9-knockout on `r genename` expression
DESeq2 was performed on all wildtype (WT) against all *klf9*-knockout (KO) samples at each time. Each run used PC1 as a covariant to underscore the effect on gene expression of the *klf9*-knockout at that time isolated from the replicate effect. Genes with a positive fold change are upregulated in *klf9* knockouts.

The plots below demonstrates differential expression of `r genename` due to the *klf9*-knockout at each time both in the context of all 24,873 genes summarized in the three analyses and alone. Low gene counts (i.e. <100) make log2FoldChange a less reliable measure of differential expression.

```{r DESEQtimespecific, echo=FALSE,warning=FALSE, out.width=c('50%','50%'), fig.show='hold'}

## violin plot of FC of all genes at the different times with specific gene as a point
#to manage the scale of the plot, cutoff at logFoldChange (-3,3) unless the min/max is lower/higher
fcmin <- -3
fcmax <- 3 
if(any(subset(long_dsq,GeneName==genename)$log2FoldChange<fcmin)){
  fcmin<- min(subset(long_dsq,GeneName==genename)$log2FoldChange)
}
if(any(subset(long_dsq,GeneName==genename)$log2FoldChange>fcmax)){
  fcmax<- max(subset(long_dsq,GeneName==genename)$log2FoldChange)
}
ggplot(data=long_dsq, aes(x=time,y=log2FoldChange))+geom_violin()+
  ylim(fcmin,fcmax)+
  geom_point(data=subset(long_dsq,GeneName==genename),aes(y=log2FoldChange),color="green",size=4)+ 
  geom_hline(yintercept=0,linetype="dashed")+
  theme(legend.position = "none", aspect.ratio = 1.5,panel.background = element_blank())+
  labs(x="Time", y= "log2 Fold Change")+
  ggtitle(paste("Effect of klf9-knockout on ",genename," expression \ncompared to all genes",sep=""))

# focus on time-dependent effect: plot of FC of this gene at different times labeled with mean count and padj
ggplot(data=subset(long_dsq,GeneName==genename), aes(x=time,y=log2FoldChange))+geom_bar(stat="identity", position=position_dodge(),fill="green")+
  facet_grid(. ~ time, scales='free',labeller=label_both)+
  geom_text(data=subset(long_dsq,GeneName==genename),aes(label = paste("padj: ",round(padj,2),"\n count: ",round(baseMean,0),sep="")),vjust="inward", size=3)+
  geom_hline(yintercept=0)+
  theme(legend.position = "none",
        aspect.ratio = 1.5,
        axis.title.x=element_blank(),
        panel.background = element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(), panel.border = element_blank(), panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())+ 
  labs(x="Time", y= "log2 Fold Change")+
  ggtitle(paste("Effect of klf9-knockout on ",genename," expression",sep=""))


```
Another run of DESEQ identified differential expression in *klf9*-knockout samples from all times vs. all wildtype samples. Both PC1 and time were used as covariants to isolate the effect of the *klf9*-knockout on gene expression across the experiment. `r genename` is highlighted in green. 

```{r MAplot, echo=FALSE,warning=FALSE}
#MA plot
ggplot(data=deseq, aes(x=log2(baseMean),y=log2FoldChange))+geom_point(size=1,color="black")+
  geom_hline(yintercept=0,linetype="dashed")+
  geom_point(data=subset(deseq,GeneName==genename),size=4,pch=18,color="green")+
  theme(legend.position = "none", aspect.ratio = 1.5,panel.background = element_blank())+
  labs(x="log2 mean transcript count (normalized for sequencing depth)", y= "klf9-knockout log2 Fold Change")+
  ggtitle(paste("MA plot: all jcoffman008 WT v. KO with ",genename," (padj= ", round(subset(deseq,GeneName==genename)$padj,2),")",  sep=""))
  ```

